### DISCUSSION TITLE
Calendar entity returns undefined

### DISCUSSION THREAD

**Discussion by pdhruv93** [shannonhochkins/ha-component-kit#57] [Q&A]
The listEvents is returning undefined. I can however see events on my calendar under HA
```
const { api } = useEntity('calendar.work_meetings');
  return <button onClick={() => {
    api.listEvents();
}}
```

**Comment by shannonhochkins** [comment]
Can you please provide calendar settings in home assistant? 

Have you tried the same service in home assistant directly?

Have you checked the documentation on the list events to see if it takes any additional required parameters?

I will take a look in the morning but more context would be helpful please

**Reply by shannonhochkins** [reply]
I also have not had a chance to play around with calendars much yet with hakit/core, I use Google calendars and hadn't seen the integration in home assistant to be that interesting myself but maybe ita because I've not learnt about them

**Reply by pdhruv93** [reply]
aah so when checking with HA developer mode, list_events needs to have startDate and endDate params. I thought these would be marked as required in Typescript.
Moreover, it asks me to pass the start and end dates as object. Not sure what format of object it is expecting

![image](https://github.com/shannonhochkins/ha-component-kit/assets/38987871/468245c4-1c62-4944-9604-6ca47949b103)

**Reply by shannonhochkins** [reply]
The types are generated from the home assistant services schema, which is not perfect unfortunately and changes almost daily, I do have some entities and services completely re written, if you look at the documentation under the typescript sync section, I have documentation on how to use module augmentation to fix the types your side, once you're happy and it's all working feel free to send me the types back and I'll integrate them!

Sorry, it was already a massive task to get the types to this level, as you can imagine they aren't perfect

**Reply by shannonhochkins** [reply]
Media player had the same issue with  the join service, says it expects a string but it accepts an array of strings for group_members

**Reply by shannonhochkins** [reply]
Also, if you come up with something nice for a calendar, send it over it may save me some time when I get to building a card for it :)

**Reply by dhruvpandeysm** [reply]
@shannonhochkins can you please provide the correct format for the start and end date, it's expecting an object.
```
.listEvents({
    start_date_time: {time: "2023-09-20"},
    end_date_time: {"2023-09-22T13:48:23.498075+00:00"},
  });
```

**Reply by shannonhochkins** [reply]
As mentioned above, if something doesn't work the way you're expecting it to, it's because the schema for the services provided by home assistant is wrong.

```
// Lists events on a calendar within a time range.
      listEvents: ServiceFunction<
        T,
        {
          // Returns active events after this time (exclusive). When not set, defaults to now. @example 2022-03-22 20:00:00
          start_date_time?: object;
          // Returns active events before this time (exclusive). Cannot be used with 'duration'. @example 2022-03-22 22:00:00
          end_date_time?: object;
          // Returns active events from start_date_time until the specified duration.
          duration?: object;
        }
      >;
      ```
      
      As you said, it's expecting an object for start/end time, but if you have a look at the description there, it's clear that it shouldnt be an object, it should be a string, have you tried as a string?

**Reply by shannonhochkins** [reply]
```
const getDatePlus20Days = (): string => {
  const currentDate = new Date();
  currentDate.setDate(currentDate.getDate() + 20);
  return currentDate.toISOString();
};

const datePlus20Days = getDatePlus20Days();
.listEvents({
    // @ts-expect-error - this is correct, types are wrong from home assistant
    end_date_time: datePlus20Days,
  });
  ```
  
  You may need to play around with the ormatting of the date string to match the description if iso format doesn't work
